crates/resolver/src/hosts.rs

```diff
@@ -47,14 +49,62 @@ impl Hosts {
     /// Look up the addresses for the given host from the system hosts file.
     pub fn lookup_static_host(&self, query: &Query) -> Option<Lookup> {
         if !self.by_name.is_empty() {
-            if let Some(val) = self.by_name.get(query.name()) {
-                let result = match query.query_type() {
-                    RecordType::A => val.a.clone(),
-                    RecordType::AAAA => val.aaaa.clone(),
-                    _ => None,
-                };
-
-                return result;
+            match query.query_type() {
+                RecordType::A | RecordType::AAAA => {
+                    if let Some(val) = self.by_name.get(query.name()) {
+                        let result = match query.query_type() {
+                            RecordType::A => val.a.clone(),
+                            RecordType::AAAA => val.aaaa.clone(),
+                            _ => None,
+                        };
+
+                        return result;
+                    }
+                }
+                RecordType::PTR => {
+                    if let Ok(ip) = query.name().parse_arpa_name() {
+                        let ip = ip.addr();
+                        let mut hostnames = self
+                            .by_name
+                            .iter()
+                            .filter(|(_, v)| match ip {
+                                IpAddr::V4(ip) => match v.a.as_ref() {
+                                    Some(lookup) => lookup.iter().any(|r| {
+                                        r.ip_addr().map(|it| it == ip).unwrap_or_default()
+                                    }),
+                                    None => false,
+                                },
+                                IpAddr::V6(ip) => match v.aaaa.as_ref() {
+                                    Some(lookup) => lookup.iter().any(|r| {
+                                        r.ip_addr().map(|it| it == ip).unwrap_or_default()
+                                    }),
+                                    None => false,
+                                },
+                            })
+                            .map(|(n, _)| n)
+                            .collect::<Vec<_>>();
+
+                        if !hostnames.is_empty() {
+                            hostnames.sort();
+                            let records = hostnames
+                                .into_iter()
+                                .map(|hostname| {
+                                    Record::from_rdata(
+                                        query.name().clone(),
+                                        dns_lru::MAX_TTL,
+                                        RData::PTR(PTR(hostname.clone())),
+                                    )
+                                })
+                                .collect::<Vec<_>>();
```

Comment:
It looks like we can chain the `map()` here onto `hostnames`, avoiding the intermediate allocation.

Reply:
I don't quite understand this, but I've adjusted it a bit. Please take a look at it again.